File: typescript/41.ts
function firstMissingPositive(nums: number[]): number {
    const n = nums.length;
    for (let i = 0; i < n; i++) {
        const num = nums[i];
        if (num > 0 && num <= n && nums[num - 1] !== num) {
            nums[i] = nums[num - 1];
            nums[num - 1] = num;
            i--;
        }
    }
    for (let i = 0; i < n; i++) {
        if (nums[i] !== i + 1) return i + 1;
    }
    return n + 1;
};

test('41', () => {
    expect(firstMissingPositive([1, 2, 0])).toBe(3)
    expect(firstMissingPositive([3, 4, -1, 1])).toBe(2)
    expect(firstMissingPositive([7, 8, 9, 11, 12])).toBe(1)
})

File: python/41.py
import unittest
from typing import List


class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n, i = len(nums), 0
        while i < n:
            num = nums[i]
            if num > 0 and num <= n and nums[num - 1] != num:
                nums[num - 1], nums[i] = nums[i], nums[num - 1]
                i -= 1
            else:
                i += 1
        for i, num in enumerate(nums):
            if i + 1 != num:
                return i + 1
        return n + 1


class TestSolution(unittest.TestCase):
    def test(self):
        solution = Solution()
        self.assertEqual(solution.firstMissingPositive([1, 2, 0]), 3)
        self.assertEqual(solution.firstMissingPositive([3, 4, -1, 1]), 2)
